File: admin/src/data/flags.ts
import { useCallback } from 'react'

import {
	bosonFamily,
	useSetBoson,
	useQuery,
	usePost,
} from '@biotic-ui/boson'

import { useHttp } from 'data/http'

export enum Flags {
	SignIn = 'auth::signin',
	SignUp = 'auth::signup',
	PasswordReset = 'action::password_reset',
	VerifyEmail = 'action::verify_email',
	
	EmailAndPassword = 'method::email_password',
	AuthenticationLink = 'method::authentication_link',

	OAuthGoogle = 'oauth::google',
}

export function isFlag(flag: string | Flags): boolean {
	let index = Object.values(Flags).findIndex(value => {
		return value === flag
	})

	return index !== -1
}

export function getFlagsFromRequest(flags: Array<string>): Array<Flags> {
	let items = flags
		.filter(flag => isFlag(flag))

	return (items as Array<Flags>)
}


type ProjectFlags = Array<Flags>

type Response = {
	items: Array<string>
}


let flagsFamily = bosonFamily<[string], ProjectFlags>(() => {
	return {
		defaultValue: []
	}
})

export function useFlags(project: string) {
	let http = useHttp()

	return useQuery(flagsFamily(project), async () => {
		let res = await http.get<Response>('project/flags',  {
			params: {
				project
			}
		})

		return getFlagsFromRequest(res.data.items)
	})
}

export function useToggleFlags(project: string) {
	let setState = useSetBoson(flagsFamily(project))
	
	let set = useCallback((flag: Flags) => () => {
		setState((items = []) => {
			if (items.includes(flag) && flag === Flags.EmailAndPassword) {
				return items.filter(f => (
					f !== Flags.EmailAndPassword &&
					f !== Flags.PasswordReset &&
					f !== Flags.VerifyEmail
				))
			}

			if (items.includes(flag)) {
				return items.filter(f => f !== flag)
			}

			return items.concat([flag])
		})
	}, [setState])

	return set
}


type UpdateFlags = {
	project: string;
	flags: Array<Flags>;
}

export function useUpdateFlags(project: string) {
	let http = useHttp()
	return usePost(async (flags: Array<Flags>) => {
		let payload: UpdateFlags = { project, flags }
		await http.post<void>('project/set_flags', payload)
	})
}

File: server/src/project/data/flags.rs
use crate::response::error::ApiError;

use serde::{Deserialize, Serialize};
use sqlx::PgPool;
use uuid::Uuid;

#[derive(Debug, Copy, Clone, Deserialize, Serialize, PartialEq)]
pub enum Flags {
    #[serde(rename = "auth::signin")]
    SignIn,
    #[serde(rename = "auth::signup")]
    SignUp,
    #[serde(rename = "action::password_reset")]
    PasswordReset,
    #[serde(rename = "action::verify_email")]
    VerifyEmail,
    #[serde(rename = "method::email_password")]
    EmailAndPassword,
    #[serde(rename = "method::authentication_link")]
    AuthenticationLink,

    #[serde(rename = "oauth::google")]
    OAuthGoogle,
}

impl Flags {
    pub async fn from_project(pool: &PgPool, id: &Uuid) -> Result<Vec<Flags>, ApiError> {
        let row = sqlx::query_file!("src/project/sql/get_flags.sql", id)
            .fetch_one(pool)
            .await
            .map_err(|_| ApiError::InternalServerError)?;

        let flags: Vec<Flags> = row
            .flags
            .iter()
            .map(|flag| Flags::from_str(&flag))
            .filter(|result| result.is_some())
            .map(|result| result.unwrap())
            .collect();

        Ok(flags)
    }

    pub async fn set_flags(pool: &PgPool, project: &Uuid, flags: &[Flags]) -> Result<(), ApiError> {
        let flags = flags
            .into_iter()
            .map(|flag| flag.to_string())
            .filter(|flag| flag.is_empty() == false)
            .collect::<Vec<String>>();

        sqlx::query_file!("src/project/sql/set_flags.sql", project, &flags)
            .execute(pool)
            .await
            .map_err(|_| ApiError::InternalServerError)?;

        Ok(())
    }

    pub async fn has_flags(
        pool: &PgPool,
        project: &Uuid,
        flags: &[Flags],
    ) -> Result<bool, ApiError> {
        let project_flags = Flags::from_project(pool, project).await?;

        let contains: Vec<bool> = flags
            .iter()
            .map(|flag| project_flags.contains(flag))
            .filter(|in_flags| *in_flags == false)
            .collect();

        if contains.len() == 0 {
            Ok(true)
        } else {
            Err(ApiError::Forbidden)
        }
    }

    // TODO figure out how to do this automatically?
    fn from_str(flag: &str) -> Option<Flags> {
        match flag {
            "auth::signin" => Some(Flags::SignIn),
            "auth::signup" => Some(Flags::SignUp),
            "action::password_reset" => Some(Flags::PasswordReset),
            "method::email_password" => Some(Flags::EmailAndPassword),
            "method::authentication_link" => Some(Flags::AuthenticationLink),
            "action::verify_email" => Some(Flags::VerifyEmail),
            "oauth::google" => Some(Flags::OAuthGoogle),
            _ => None,
        }
    }
}

impl ToString for Flags {
    fn to_string(&self) -> String {
        match self {
            Flags::SignIn => "auth::signin".to_string(),
            Flags::SignUp => "auth::signup".to_string(),
            Flags::PasswordReset => "action::password_reset".to_string(),
            Flags::EmailAndPassword => "method::email_password".to_string(),
            Flags::AuthenticationLink => "method::authentication_link".to_string(),
            Flags::VerifyEmail => "action::verify_email".to_string(),
            Flags::OAuthGoogle => "oauth::google".to_string(),
        }
    }
}
